In [2]:
import pandas as pd
import numpy as np


import time, warnings
import datetime as dt

#modules for predictive models
import sklearn.cluster as cluster
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
#from sklearn.mixture import GMM

from sklearn.metrics import silhouette_samples, silhouette_score

#visualizations
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
%matplotlib inline
import seaborn as sns

warnings.filterwarnings("ignore")

In [3]:
retail_df = pd.read_excel('C:/Users/ITQLAP/Desktop/garbage/Online Retail.xlsx')
retail_df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [4]:
retail_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [5]:
retail_df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

As customer clusters may vary by geography, I’ll restrict the data to only United Kingdom customers, which contains most of our customers historical data.

In [6]:
retail_uk = retail_df[retail_df['Country']=='United Kingdom']
#check the shape
retail_uk.shape

(495478, 8)

In [36]:
retail_uk['Quantity'].nunique()

180

In [8]:
#remove canceled orders
retail_uk = retail_uk[retail_uk['Quantity']>0]
retail_uk.shape

(486286, 8)

In [9]:
retail_uk['CustomerID'].isnull().sum()

131941

In [10]:
#remove the row where customer ID are NA
retail_uk.dropna(subset=['CustomerID'],how='all',inplace = True)
retail_uk.shape

(354345, 8)

In [11]:
retail_uk['InvoiceDate'].nunique()

15615

In [12]:
#restrict the data to one full year because it's better to use a metric per Months or Years in RFM
retail_uk = retail_uk[retail_uk['InvoiceDate']>= "2010-12-09"]
retail_uk.shape

(342478, 8)

In [13]:
print("Summary..")
#exploring the unique values of each attribute
print("Number of transactions: ", retail_uk['InvoiceNo'].nunique())
print("Number of products bought: ",retail_uk['StockCode'].nunique())
print("Number of customers:", retail_uk['CustomerID'].nunique() )
print("Percentage of customers NA: ", round(retail_uk['CustomerID'].isnull().sum() * 100 / len(retail_df),2),"%" )
# 

Summary..
Number of transactions:  16017
Number of products bought:  3611
Number of customers: 3863
Percentage of customers NA:  0.0 %


# RFM analysis

RFM (Recency, Frequency, Monetary) analysis is a customer segmentation technique that uses past purchase behavior to divide customers into groups.
RFM helps divide customers into various categories or clusters to identify customers who are more likely to respond to promotions and also for future personalization services.

* RECENCY (R): Days since last purchase
* FREQUENCY (F): Total number of purchases
* MONETARY VALUE (M): Total money this customer spent.

We will create those 3 customer attributes for each customer.

# RECENCY Analysis

In [14]:
# First date and last date available in our dataset
print('First Date in Datasets : ', retail_uk['InvoiceDate'].min())
print('Last date in Datasets : ',retail_uk['InvoiceDate'].max())

First Date in Datasets :  2010-12-09 08:34:00
Last date in Datasets :  2011-12-09 12:49:00


In [15]:
now = dt.date(2011,12,9)
print(now)

2011-12-09


In [16]:
#create a new column called date which contains the date of invoice only
retail_uk['date'] = retail_uk['InvoiceDate'].dt.date

In [17]:
retail_uk.head()

InvoiceNo StockCode                          Description  Quantity  \
19632    537879     22114    HOT WATER BOTTLE TEA AND SYMPATHY        12   
19633    537879     22835      HOT WATER BOTTLE I AM SO POORLY         8   
19634    537879     85150        LADIES & GENTLEMEN METAL SIGN         6   
19635    537879     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS         4   
19636    537879     21524       DOORMAT SPOTTY HOME SWEET HOME         2   

              InvoiceDate  UnitPrice  CustomerID         Country        date  
19632 2010-12-09 08:34:00       3.95     14243.0  United Kingdom  2010-12-09  
19633 2010-12-09 08:34:00       4.65     14243.0  United Kingdom  2010-12-09  
19634 2010-12-09 08:34:00       2.55     14243.0  United Kingdom  2010-12-09  
19635 2010-12-09 08:34:00       7.95     14243.0  United Kingdom  2010-12-09  
19636 2010-12-09 08:34:00       7.95     14243.0  United Kingdom  2010-12-09

In [18]:
#group by customers and check last date of purshace
recency_df = retail_uk.groupby(by='CustomerID', as_index=False)['date'].max()
recency_df.columns = ['CustomerID','LastPurshaceDate']
recency_df.tail()

CustomerID LastPurshaceDate
3858     18280.0       2011-03-07
3859     18281.0       2011-06-12
3860     18282.0       2011-12-02
3861     18283.0       2011-12-06
3862     18287.0       2011-10-28

In [19]:
#calculate recency
recency_df['Recency'] = recency_df['LastPurshaceDate'].apply(lambda x: (now - x).days)

In [20]:
recency_df.head()

CustomerID LastPurshaceDate  Recency
0     12346.0       2011-01-18      325
1     12747.0       2011-12-07        2
2     12748.0       2011-12-09        0
3     12749.0       2011-12-06        3
4     12820.0       2011-12-06        3

In [21]:
#drop LastPurchaseDate as we don't need it anymore
recency_df.drop('LastPurshaceDate',axis=1,inplace=True)

Now we have the recency attribute created. e.g: Customer with ID = 12346 did his/her last purshace 325 days ago.

# Frequency

In [22]:
# copy the data
retail_uk_copy = retail_uk
retail_uk_copy.head()

InvoiceNo StockCode                          Description  Quantity  \
19632    537879     22114    HOT WATER BOTTLE TEA AND SYMPATHY        12   
19633    537879     22835      HOT WATER BOTTLE I AM SO POORLY         8   
19634    537879     85150        LADIES & GENTLEMEN METAL SIGN         6   
19635    537879     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS         4   
19636    537879     21524       DOORMAT SPOTTY HOME SWEET HOME         2   

              InvoiceDate  UnitPrice  CustomerID         Country        date  
19632 2010-12-09 08:34:00       3.95     14243.0  United Kingdom  2010-12-09  
19633 2010-12-09 08:34:00       4.65     14243.0  United Kingdom  2010-12-09  
19634 2010-12-09 08:34:00       2.55     14243.0  United Kingdom  2010-12-09  
19635 2010-12-09 08:34:00       7.95     14243.0  United Kingdom  2010-12-09  
19636 2010-12-09 08:34:00       7.95     14243.0  United Kingdom  2010-12-09

In [23]:
# drop duplicates
retail_uk_copy.drop_duplicates(subset=['InvoiceNo', 'CustomerID'], keep="first", inplace=True)

#calculate frequency of purchases
frequency_df = retail_uk_copy.groupby(by=['CustomerID'], as_index=False)['InvoiceNo'].count()
frequency_df.columns = ['CustomerID','Frequency']
frequency_df.head()

CustomerID  Frequency
0     12346.0          1
1     12747.0         10
2     12748.0        196
3     12749.0          5
4     12820.0          4

# Monetary

Monetary attribute answers the question: How much money did the customer spent over time?

To do that, first, we will create a new column total cost to have the total price per invoice.

In [24]:
retail_uk.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country', 'date'],
      dtype='object')

In [25]:
# Create a column with total cost
retail_uk['TotalCost']  = retail_uk['Quantity']*retail_uk['UnitPrice']
retail_uk.head()

InvoiceNo StockCode                        Description  Quantity  \
19632    537879     22114  HOT WATER BOTTLE TEA AND SYMPATHY        12   
19640    537880     21669    BLUE STRIPE CERAMIC DRAWER KNOB        12   
19656    537881     22730         ALARM CLOCK BAKELIKE IVORY        16   
19668    537882     22804    CANDLEHOLDER PINK HANGING HEART        32   
19672    537883     84946    ANTIQUE SILVER TEA GLASS ETCHED        12   

              InvoiceDate  UnitPrice  CustomerID         Country        date  \
19632 2010-12-09 08:34:00       3.95     14243.0  United Kingdom  2010-12-09   
19640 2010-12-09 09:14:00       1.25     12963.0  United Kingdom  2010-12-09   
19656 2010-12-09 09:23:00       3.75     15513.0  United Kingdom  2010-12-09   
19668 2010-12-09 09:25:00       2.55     15079.0  United Kingdom  2010-12-09   
19672 2010-12-09 09:27:00       1.25     14437.0  United Kingdom  2010-12-09   

       TotalCost  
19632       47.4  
19640       15.0  
19656       60.0  
19668       81.6  
19672       15.0

In [26]:
# groupby cusomer id with total cost
monetary_df = retail_uk.groupby(by='CustomerID').agg({'TotalCost':'sum'})
monetary_df.head()

TotalCost
CustomerID           
12346.0      77183.60
12747.0        658.89
12748.0       3739.23
12749.0         98.35
12820.0         58.20

In [27]:
# using of as_index=False parameters
monetary_df = retail_uk.groupby(by='CustomerID',as_index=False).agg({'TotalCost':'sum'})
monetary_df.columns = ['CustomerID','Monetary']
monetary_df.head()

CustomerID  Monetary
0     12346.0  77183.60
1     12747.0    658.89
2     12748.0   3739.23
3     12749.0     98.35
4     12820.0     58.20

# Create RFM Table

In [28]:
# Merge the dataFrame
temp_df = recency_df.merge(frequency_df,on = 'CustomerID')
temp_df.head()

CustomerID  Recency  Frequency
0     12346.0      325          1
1     12747.0        2         10
2     12748.0        0        196
3     12749.0        3          5
4     12820.0        3          4

In [29]:
#merge monetary in RFM table
rfm_df = temp_df.merge(monetary_df,on = 'CustomerID')
rfm_df.head()

CustomerID  Recency  Frequency  Monetary
0     12346.0      325          1  77183.60
1     12747.0        2         10    658.89
2     12748.0        0        196   3739.23
3     12749.0        3          5     98.35
4     12820.0        3          4     58.20

In [30]:
rfm_df.set_index('CustomerID',inplace = True)
rfm_df.head()

Recency  Frequency  Monetary
CustomerID                              
12346.0         325          1  77183.60
12747.0           2         10    658.89
12748.0           0        196   3739.23
12749.0           3          5     98.35
12820.0           3          4     58.20

# RFM table correctness verification

In [31]:
retail_uk[retail_uk['CustomerID']==12346.0]

InvoiceNo StockCode                     Description  Quantity  \
61619    541431     23166  MEDIUM CERAMIC TOP STORAGE JAR     74215   

              InvoiceDate  UnitPrice  CustomerID         Country        date  \
61619 2011-01-18 10:01:00       1.04     12346.0  United Kingdom  2011-01-18   

       TotalCost  
61619    77183.6

In [32]:
(now - dt.date(2011,1,18)).days

325

<center><h1><em>Customer Segmentation using RFM</em></h1></center>

Before moving to customer segments, Let's see the application of Pareto Principle – commonly referred to as the 80-20 rule on our dataset by applying it to our RFM variables.

Pareto’s rule says 80% of the results come from 20% of the causes.

Similarly, 20% customers contribute to 80% of your total revenue. Let's verify that because that will help us know which customers to focus on when marketing new products.

# Applying 80-20 Rule

In [33]:
rfm_df.columns

Index(['Recency', 'Frequency', 'Monetary'], dtype='object')

In [34]:
# Get 80% of revenue
pareto_cutoff = rfm_df['Monetary'].sum()*0.80
print("80% of total revenue :",pareto_cutoff)

80% of total revenue : 890679.536


In [37]:
# Create a new column that is rank of the value of the coverage in ascending order
customer_rank = rfm_df
customer_rank['Rank'] = customer_rank['Monetary'].rank(ascending = 0)
customer_rank.head()

Recency  Frequency  Monetary    Rank
CustomerID                                      
12346.0         325          1  77183.60     2.0
12747.0           2         10    658.89   183.0
12748.0           0        196   3739.23    34.0
12749.0           3          5     98.35  1140.0
12820.0           3          4     58.20  1665.5

# Top Rankers

In [38]:
customer_rank.sort_values(by =['Rank'],ascending = True)

Recency  Frequency   Monetary    Rank
CustomerID                                       
16446.0           0          2  168471.25     1.0
12346.0         325          1   77183.60     2.0
15098.0         182          3   39916.50     3.0
18102.0           0         57   39646.45     4.0
17949.0           1         44   28685.69     5.0
...             ...        ...        ...     ...
15503.0         362          1       0.42  3859.5
15645.0          18          1       0.42  3859.5
17914.0           3          1       0.39  3861.5
13271.0          37          1       0.39  3861.5
13256.0          14          1       0.00  3863.0

[3863 rows x 4 columns]

# Get 20% of the customer

In [40]:
top_20_cutoff = 3863*0.20
top_20_cutoff

772.6

In [41]:
# sum the monetary values over the customer with rank <=773
revenueByTop20 = customer_rank[customer_rank['Rank']<=772]['Monetary'].sum()
revenueByTop20

976683.3499999999

* 80% of total revenue : 890679.536
* but in actual revenue should be 976683
* Hence, the 80% of total revenue is not achieved by the 20% of TOP customers but approximately, it does, because they are less than our 20% TOP customers who achieve it. It would be interesting to study this group of customers because they are those who make our most revenue.

# applying RFM score formula

The simplest way to create customers segments from RFM Model is to use Quartiles. We assign a score from 1 to 4 to Recency, Frequency and Monetary. Four is the best/highest value, and one is the lowest/worst value. A final RFM score is calculated simply by combining individual RFM score numbers.

Note: Quintiles (score from 1-5) offer better granularity, in case the business needs that but it will be more challenging to create segments since we will have 555 possible combinations. So, we will use quartiles.

DataFrame.quantile(q=0.5, axis=0, numeric_only=True, interpolation='linear')

In [45]:
quantiles = rfm_df.quantile(q= [0.25,0.5,0.75],numeric_only = True)
quantiles

Recency  Frequency  Monetary    Rank
0.25     17.0        1.0      17.4   966.5
0.50     49.0        2.0      45.0  1930.5
0.75    134.0        5.0     121.6  2898.0

In [47]:
# convert data Frame to dict
quantiles.to_dict()

{'Recency': {0.25: 17.0, 0.5: 49.0, 0.75: 134.0},
 'Frequency': {0.25: 1.0, 0.5: 2.0, 0.75: 5.0},
 'Monetary': {0.25: 17.4, 0.5: 45.0, 0.75: 121.6},
 'Rank': {0.25: 966.5, 0.5: 1930.5, 0.75: 2898.0}}

# Create RFM segmentation table

We will create two segmentation classes since, high recency is bad, while high frequency and monetary value is good.

In [56]:
# Arguments (x = value, p = recency, monetary_value, frequency, d = quartiles dict)
def RScore(x,p,d):
    if x <= d[p][0.25]:
        return 4
    elif x <= d[p][0.50]:
        return 3
    elif x <= d[p][0.75]: 
        return 2
    else:
        return 1
# Arguments (x = value, p = recency, monetary_value, frequency, k = quartiles dict)
def FMScore(x,p,d):
    if x <= d[p][0.25]:
        return 1
    elif x <= d[p][0.50]:
        return 2
    elif x <= d[p][0.75]: 
        return 3
    else:
        return 4

In [54]:
# Create rfm segmentation table
rfm_seg =rfm_df
rfm_seg['R_Quartile'] = rfm_seg['Recency'].apply(RScore,args=('Recency',quantiles))
rfm_seg['F_Quartile'] = rfm_seg['Frequency'].apply(FMScore,args=('Frequency',quantiles))
rfm_seg['M_Quartile'] = rfm_seg['Monetary'].apply(FMScore,args=('Monetary',quantiles))

In [55]:
rfm_seg.head()

Recency  Frequency  Monetary    Rank  R_Quartile  F_Quartile  \
CustomerID                                                                 
12346.0         325          1  77183.60     2.0           1           1   
12747.0           2         10    658.89   183.0           4           4   
12748.0           0        196   3739.23    34.0           4           4   
12749.0           3          5     98.35  1140.0           4           3   
12820.0           3          4     58.20  1665.5           4           3   

            M_Quartile  
CustomerID              
12346.0              4  
12747.0              4  
12748.0              4  
12749.0              3  
12820.0              3

First, we need to combine the scores (R_Quartile, F_Quartile,M_Quartile) together.

In [58]:
rfm_seg['RFMScore'] = rfm_seg.R_Quartile.map(str) \
                            + rfm_seg.F_Quartile.map(str) \
                            + rfm_seg.M_Quartile.map(str)
rfm_seg.head()

Recency  Frequency  Monetary    Rank  R_Quartile  F_Quartile  \
CustomerID                                                                 
12346.0         325          1  77183.60     2.0           1           1   
12747.0           2         10    658.89   183.0           4           4   
12748.0           0        196   3739.23    34.0           4           4   
12749.0           3          5     98.35  1140.0           4           3   
12820.0           3          4     58.20  1665.5           4           3   

            M_Quartile RFMScore  
CustomerID                       
12346.0              4      114  
12747.0              4      444  
12748.0              4      444  
12749.0              3      433  
12820.0              3      433

Best Recency score = 4: most recently purchase. Best Frequency score = 4: most quantity purchase. Best Monetary score = 4: spent the most.

Let's see who are our **Champions** (best customers).

In [60]:
rfm_seg[rfm_seg['RFMScore']=='444'].sort_values('Monetary', ascending=False).head(10)

Recency  Frequency  Monetary  Rank  R_Quartile  F_Quartile  \
CustomerID                                                               
18102.0           0         57  39646.45   4.0           4           4   
17949.0           1         44  28685.69   5.0           4           4   
17450.0           8         44  25953.51   6.0           4           4   
16013.0           3         45  17163.28   8.0           4           4   
16333.0           7         22  14418.96   9.0           4           4   
15769.0           7         25  11660.84  11.0           4           4   
12901.0           8         28   9230.45  12.0           4           4   
13798.0           1         56   7786.26  13.0           4           4   
16684.0           4         28   7700.08  14.0           4           4   
17857.0           4         23   7655.18  15.0           4           4   

            M_Quartile RFMScore  
CustomerID                       
18102.0              4      444  
17949.0              4      444  
17450.0              4      444  
16013.0              4      444  
16333.0              4      444  
15769.0              4      444  
12901.0              4      444  
13798.0              4      444  
16684.0              4      444  
17857.0              4      444

**How many customers do we have in each segment?**

In [61]:
print("Best Customers: ",len(rfm_seg[rfm_seg['RFMScore']=='444']))
print('Loyal Customers: ',len(rfm_seg[rfm_seg['F_Quartile']==4]))
print("Big Spenders: ",len(rfm_seg[rfm_seg['M_Quartile']==4]))
print('Almost Lost: ', len(rfm_seg[rfm_seg['RFMScore']=='244']))
print('Lost Customers: ',len(rfm_seg[rfm_seg['RFMScore']=='144']))
print('Lost Cheap Customers: ',len(rfm_seg[rfm_seg['RFMScore']=='111']))

Best Customers:  356
Loyal Customers:  752
Big Spenders:  966
Almost Lost:  64
Lost Customers:  9
Lost Cheap Customers:  353


<h1>Conclusion - perspective from this level of customer segmentation</h1>

To gain even further insight into customer behavior, we can dig deeper in the relationship between RFM variables.

RFM model can be used in conjunction with certain predictive models like k-means clustering, Logistic Regression and Recommendation to produce better informative results on customer behavior.

We will go for k-means since it has been widely used for Market Segmentation and it offers the advantage of being simple to implement, following Andrew Ng who advice in his Machine Learning course, start with a dirty and simple model then move to more complex models because simple implementation helps having a first glance at the data and know where/how to exploit it better.

<center><h1>THE END</h1></center>